src/cardano_fmt_plug.c

```diff
@@ -0,0 +1,244 @@
+#endif
+#endif
+#define BENCHMARK_COMMENT   ""
+#define BENCHMARK_LENGTH    0x107
+#define BINARY_SIZE         0
+#define BINARY_ALIGN        1
+#define SALT_SIZE           sizeof(struct custom_salt)
+#define SALT_ALIGN          sizeof(uint32_t)
+#ifdef SIMD_COEF_64
+#define MIN_KEYS_PER_CRYPT  SSE_GROUP_SZ_SHA512
+#define MAX_KEYS_PER_CRYPT  (SSE_GROUP_SZ_SHA512 * 4)
+#else
+#define MIN_KEYS_PER_CRYPT  1
+#define MAX_KEYS_PER_CRYPT  4
+#endif
+#if !defined(PLAINTEXT_LENGTH) || PLAINTEXT_LENGTH > 15
+#undef PLAINTEXT_LENGTH
+#define PLAINTEXT_LENGTH    15
+#endif
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int any_cracked, *cracked;
+static size_t cracked_size;
+
+static struct custom_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	any_cracked = 0;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void cardano_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+
+	if (any_cracked) {
+		memset(cracked, 0, cracked_size);
+		any_cracked = 0;
+	}
+	// KDF2 params
+	static const int kdf_rounds = 15000;
+	static const char kdf_salt[] = "encrypted wallet salt";
+	static int kdf_salt_len = sizeof(kdf_salt);
+
+	// ChaCha20 params
+	static const int chacha_rounds = 20;
+
+#ifdef _OPENMP
+	#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
+		int i;
+
+		unsigned char kdf_out[MIN_KEYS_PER_CRYPT][BUF_SIZE];
+		unsigned char blake_out[MIN_KEYS_PER_CRYPT][PWD_HASH_LEN];
+
+#if SIMD_COEF_64
+		int lens[MIN_KEYS_PER_CRYPT];
+		unsigned char *pin[MIN_KEYS_PER_CRYPT]; // blake2b_256 hashed password
+		unsigned char *pout[MIN_KEYS_PER_CRYPT]; // 40-byte length KDF
+#endif
+
+		for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
+			// blake2b_256 hash of the password.
+			blake2b(blake_out[i], (unsigned char *)saved_key[index + i], NULL, PWD_HASH_LEN,
+			        strlen(saved_key[index + i]), 0);
+#if SIMD_COEF_64
+			lens[i] = PWD_HASH_LEN;
+			pin[i] = (unsigned char *)blake_out[i];
+			pout[i] = (unsigned char *)kdf_out[i];
+		}
+
+		pbkdf2_sha512_sse((const unsigned char **)pin, lens, (unsigned char *)kdf_salt, kdf_salt_len, kdf_rounds, pout, BUF_SIZE, 0);
+#else
+			pbkdf2_sha512(blake_out[i], PWD_HASH_LEN,
+			              (unsigned char *)kdf_salt, kdf_salt_len, kdf_rounds, kdf_out[i], BUF_SIZE, 0);
+		}
+#endif
+
+		ed25519ext_secret_key dsks[MIN_KEYS_PER_CRYPT];
```

Comment:
Looks like the `dsks` array can be replaced with one `dsk` - only one element is used at a time. It's the same as `pk` and `ckey` in that respect - so for consistency don't use arrays for any.

`kdf_out` is different and actually needs to stay an array, because it's produced by a separate loop that can use SIMD, thus computing multiple KDF values concurrently.

Reply:
Yeah, you're right, fixed.